# Predicting sentiment from product reviews

# Fire up GraphLab Create

In [82]:
import graphlab

# Read some product review data

Loading reviews for a set of baby products. 

In [83]:
products = graphlab.SFrame('amazon_baby.gl/')

# Let's explore this data together

Data includes the product name, the review text and the rating of the review. 

In [4]:
products.head()

Columns:
	name	str
	review	str
	rating	float

Rows: 10

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|    Planetwise Flannel Wipes   | These flannel wipes are OK... |  3.0   |
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
+-------------------------------+-------------------------------+--------+
[10 rows x 3 columns]

# Build the word count vector for each review

In [84]:
products['word_count'] = graphlab.text_analytics.count_words(products['review'])

In [85]:
products.head()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict

Rows: 10

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|    Planetwise Flannel Wipes   | These flannel wipes are OK... |  3.0   |
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+
|           word_count          |
+-------------------------------+
| {'and': 5, '6': 1, 'stink'... |
| {'and': 3, 'love': 1, 'it'... |
| {'and': 2, 'quilt': 1, 'it... |
| {'ingenious': 1, 'and': 3,... |
| {'and': 2, 'parents!!': 1,... |
| {'and': 2, 'this': 2, 'her... |
| {'shop': 1, 'noble': 1, 'i... |
| {'and': 2, 'all': 1, 'righ... |
| {'and': 1, 'help': 1, 'giv... |
| {'journal.': 1, 'nanny': 1... |
+-------------------------------+
[10 rows x 4 columns]

In [86]:
graphlab.canvas.set_target('ipynb')

In [87]:
products['name'].show()

# Examining the reviews for most-sold product:  'Vulli Sophie the Giraffe Teether'

In [13]:
giraffe_reviews = products[products['name'] == 'Vulli Sophie the Giraffe Teether']

In [14]:
len(giraffe_reviews)

785

In [11]:
giraffe_reviews['rating'].show(view='Categorical')

# Build a sentiment classifier

In [12]:
products['rating'].show(view='Categorical')

## Define what's a positive and a negative sentiment

We will ignore all reviews with rating = 3, since they tend to have a neutral sentiment.  Reviews with a rating of 4 or higher will be considered positive, while the ones with rating of 2 or lower will have a negative sentiment.   

In [88]:
#ignore all 3* reviews
products = products[products['rating'] != 3]

In [89]:
#positive sentiment = 4* or 5* reviews
products['sentiment'] = products['rating'] >=4

In [91]:
products.head()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int

Rows: 10

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
| Baby Tracker&reg; - Daily ... | I love this journal and ou... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-----------+
|           word_count          | sentiment |
+-------------------------------+-----------+
| {'and': 3, 'love': 1, 'it'... |     1     |
| {'and': 2, 'quilt': 1, 'it... |     1     |
| {'ingenious': 1, 'and': 3,... |     1     |
| {'and': 2, 'parents!!': 1,... |     1     |
| {'and': 2, 'this': 2, 'her... |     1     |
| {'shop': 1, 'noble': 1, 'i... |     1     |
| {'and': 2, 'all': 1, 'righ... |     1     |
| {'and': 1, 'help': 1, 'giv... |     1     |
| {'journal.': 1, 'nanny': 1... |     1     |
| {'all': 1, 'forget': 1, 'j... |     1     |
+-------------------------------+-----------+
[10 rows x 5 columns]

## Let's train the sentiment classifier

In [17]:
train_data,test_data = products.random_split(.8, seed=0)

In [18]:
sentiment_model = graphlab.logistic_classifier.create(train_data,
                                                     target='sentiment',
                                                     features=['word_count'],
                                                     validation_set=test_data)

PROGRESS: Logistic regression:
PROGRESS: --------------------------------------------------------
PROGRESS: Number of examples          : 133448
PROGRESS: Number of classes           : 2
PROGRESS: Number of feature columns   : 1
PROGRESS: Number of unpacked features : 219217
PROGRESS: Number of coefficients    : 219218
PROGRESS: Starting L-BFGS
PROGRESS: --------------------------------------------------------
PROGRESS: +-----------+----------+-----------+--------------+-------------------+---------------------+
PROGRESS: | Iteration | Passes   | Step size | Elapsed Time | Training-accuracy | Validation-accuracy |
PROGRESS: +-----------+----------+-----------+--------------+-------------------+---------------------+
PROGRESS: | 1         | 5        | 0.000002  | 3.087719     | 0.841481          | 0.839989            |
PROGRESS: | 2         | 9        | 3.000000  | 5.161244     | 0.947425          | 0.894877            |
PROGRESS: | 3         | 10       | 3.000000  | 5.933105     | 0.92

# Evaluate the sentiment model

In [22]:
sentiment_model.evaluate(test_data, metric='roc_curve')

{'roc_curve': Columns:
 	threshold	float
 	fpr	float
 	tpr	float
 	p	int
 	n	int
 
 Rows: 1001
 
 Data:
 +------------------+----------------+------------------+-------+------+
 |    threshold     |      fpr       |       tpr        |   p   |  n   |
 +------------------+----------------+------------------+-------+------+
 |       0.0        | 0.221473999252 | 0.00513936971341 | 28019 | 5346 |
 | 0.0010000000475  | 0.778526000748 |  0.994860630287  | 28019 | 5346 |
 | 0.00200000009499 | 0.739618406285 |  0.993647167993  | 28019 | 5346 |
 | 0.00300000002608 | 0.717171717172 |  0.993004746779  | 28019 | 5346 |
 | 0.00400000018999 | 0.701833146278 |  0.992469395767  | 28019 | 5346 |
 | 0.00499999988824 | 0.690796857464 |  0.992076805025  | 28019 | 5346 |
 | 0.00600000005215 | 0.681069958848 |  0.991577144081  | 28019 | 5346 |
 | 0.00700000021607 | 0.670594837262 |  0.991148863271  | 28019 | 5346 |
 | 0.00800000037998 | 0.660680882903 |  0.990756272529  | 28019 | 5346 |
 | 0.00899999961257 

In [23]:
sentiment_model.show(view='Evaluation')

# Applying the learned model to understand sentiment for Giraffe

In [22]:
giraffe_reviews['predicted_sentiment'] = sentiment_model.predict(giraffe_reviews, output_type='probability')

In [23]:
giraffe_reviews.head()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	predicted_sentiment	float

Rows: 10

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
| Vulli Sophie the Giraffe T... | He likes chewing on all th... |  5.0   |
| Vulli Sophie the Giraffe T... | My son loves this toy and ... |  5.0   |
| Vulli Sophie the Giraffe T... | There really should be a l... |  1.0   |
| Vulli Sophie the Giraffe T... | All the moms in my moms' g... |  5.0   |
| Vulli Sophie the Giraffe T... | I was a little skeptical o... |  5.0   |
| Vulli Sophie the Giraffe T... | I have been reading about ... |  5.0   |
| Vulli Sophie the Giraffe T... | My neice loves her sophie ... |  5.0   |
| Vulli Sophie the Giraffe T... | What a friendly face!  And... |  5.0   |
| Vulli Sophie the Giraffe T... | We got this just for my so... |  5.0   |
| Vulli Sophie the Giraffe T... | My baby seems to like this... |  3.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+---------------------+
|           word_count          | predicted_sentiment |
+-------------------------------+---------------------+
| {'and': 1, 'all': 1, 'beca... |    0.999513023521   |
| {'and': 1, 'right': 1, 'he... |    0.999320678306   |
| {'and': 2, 'all': 1, 'woul... |    0.013558811687   |
| {'and': 2, 'one!': 1, 'all... |    0.995769474148   |
| {'and': 3, 'all': 1, 'mont... |    0.662374415673   |
| {'and': 6, 'seven': 1, 'al... |    0.999997148186   |
| {'and': 4, 'drooling,': 1,... |    0.989190989536   |
| {'and': 3, 'chew': 1, 'be'... |    0.999563518413   |
| {'chew': 2, 'seemed': 1, '... |    0.970160542725   |
| {'and': 2, 'already': 1, '... |    0.195367644588   |
+-------------------------------+---------------------+
[10 rows x 5 columns]

## Sort the reviews based on the predicted sentiment and explore

In [24]:
giraffe_reviews = giraffe_reviews.sort('predicted_sentiment', ascending=False)

In [25]:
giraffe_reviews.head()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	predicted_sentiment	float

Rows: 10

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
| Vulli Sophie the Giraffe T... | Sophie, oh Sophie, your ti... |  5.0   |
| Vulli Sophie the Giraffe T... | I'm not sure why Sophie is... |  4.0   |
| Vulli Sophie the Giraffe T... | I'll be honest...I bought ... |  4.0   |
| Vulli Sophie the Giraffe T... | We got this little giraffe... |  5.0   |
| Vulli Sophie the Giraffe T... | As a mother of 16month old... |  5.0   |
| Vulli Sophie the Giraffe T... | Sophie the Giraffe is the ... |  5.0   |
| Vulli Sophie the Giraffe T... | Sophie la giraffe is absol... |  5.0   |
| Vulli Sophie the Giraffe T... | My 5-mos old son took to t... |  5.0   |
| Vulli Sophie the Giraffe T... | My nephews and my four kid... |  5.0   |
| Vulli Sophie the Giraffe T... | Never thought I'd see my s... |  5.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+---------------------+
|           word_count          | predicted_sentiment |
+-------------------------------+---------------------+
| {'giggles': 1, 'all': 1, "... |         1.0         |
| {'adoring': 1, 'find': 1, ... |    0.999999999703   |
| {'all': 2, 'discovered': 1... |    0.999999999392   |
| {'all': 2, "don't": 1, '(l... |    0.99999999919    |
| {'cute': 1, 'all': 1, 'rev... |    0.999999998657   |
| {'just': 2, 'both': 1, 'mo... |    0.999999997108   |
| {'and': 5, 'the': 1, 'all'... |    0.999999995589   |
| {'just': 1, 'shape': 2, 'm... |    0.999999995573   |
| {'and': 4, 'chew': 1, 'all... |    0.999999989527   |
| {'giggles': 1, 'all': 1, '... |    0.999999985069   |
+-------------------------------+---------------------+
[10 rows x 5 columns]

## Most positive reviews for the giraffe

In [26]:
giraffe_reviews[0]['review']

"Sophie, oh Sophie, your time has come. My granddaughter, Violet is 5 months old and starting to teeth. What joy little Sophie brings to Violet. Sophie is made of a very pliable rubber that is sturdy but not tough. It is quite easy for Violet to twist Sophie into unheard of positions to get Sophie into her mouth. The little nose and hooves fit perfectly into small mouths, and the drooling has purpose. The paint on Sophie is food quality.Sophie was born in 1961 in France. The maker had wondered why there was nothing available for babies and made Sophie from the finest rubber, phthalate-free on St Sophie's Day, thus the name was born. Since that time millions of Sophie's populate the world. She is soft and for babies little hands easy to grasp. Violet especially loves the bumpy head and horns of Sophie. Sophie has a long neck that easy to grasp and twist. She has lovely, sizable spots that attract Violet's attention. Sophie has happy little squeaks that bring squeals of delight from Viol

In [27]:
giraffe_reviews[1]['review']

"I'm not sure why Sophie is such a hit with the little ones, but my 7 month old baby girl is one of her adoring fans.  The rubber is softer and more pleasant to handle, and my daughter has enjoyed chewing on her legs and the nubs on her head even before she started teething.  She also loves the squeak that Sophie makes when you squeeze her.  Not sure what it is but if Sophie is amongst a pile of her other toys, my daughter will more often than not reach for Sophie.  And I have the peace of mind of knowing that only edible and safe paints and materials have been used to make Sophie, as opposed to Bright Starts and other baby toys made in China.  Now that the research is out on phthalates and other toxic substances in baby toys, I think it's more important than ever to find good quality toys that are also safe for our babies to handle and put in their mouths.  Sophie is a must-have for every new mom in my opinion.  Even if your kid is one of the few that can take or leave her, it's worth

## Show most negative reviews for giraffe

In [29]:
giraffe_reviews[-1]['review']

"My son (now 2.5) LOVED his Sophie, and I bought one for every baby shower I've gone to. Now, my daughter (6 months) just today nearly choked on it and I will never give it to her again. Had I not been within hearing range it could have been fatal. The strange sound she was making caught my attention and when I went to her and found the front curved leg shoved well down her throat and her face a purply/blue I panicked. I pulled it out and she vomited all over the carpet before screaming her head off. I can't believe how my opinion of this toy has changed from a must-have to a must-not-use. Please don't disregard any of the choking hazard comments, they are not over exaggerated!"

In [31]:
giraffe_reviews[-2]['review']

"This children's toy is nostalgic and very cute. However, there is a distinct rubber smell and a very odd taste, yes I tried it, that my baby did not enjoy. Also, if it is soiled it is extremely difficult to clean as the rubber is a kind of porus material and does not clean well. The final thing is the squeaking device inside which stopped working after the first couple of days. I returned this item feeling I had overpaid for a toy that was defective and did not meet my expectations. Please do not be swayed by the cute packaging and hype surounding it as I was. One more thing, I was given a full refund from Amazon without any problem."

# Assignment

In the IPython notebook above, we used the word counts for all words in the reviews to train the sentiment classifier model. Now, we are going to follow a similar path, but only use this subset of the words:

In [92]:
selected_words = ['awesome', 'great', 'fantastic', 'amazing', 'love', 'horrible', 'bad', 'terrible', 'awful', 'wow', 'hate']

In [93]:
def word_count(word, dict):
    if word in dict:
        return dict[word]
    else:
        return 0

In [94]:
products['awesome']=products['word_count'].apply(lambda x: fn_word_count('awesome',x))

In [95]:
products['great']=products['word_count'].apply(lambda x: fn_word_count('great',x))

In [96]:
products['fantastic'] = products['word_count'].apply(lambda x: fn_word_count('fantastic',x))

In [97]:
products['amazing'] = products['word_count'].apply(lambda x: fn_word_count('amazing',x))

In [98]:
products['love'] = products['word_count'].apply(lambda x: fn_word_count('love',x))

In [99]:
products['horrible'] = products['word_count'].apply(lambda x: fn_word_count('horrible',x))

In [100]:
products['bad'] = products['word_count'].apply(lambda x: fn_word_count('bad',x))

In [101]:
products['terrible'] = products['word_count'].apply(lambda x: fn_word_count('terrible',x))

In [102]:
products['awful'] = products['word_count'].apply(lambda x: fn_word_count('awful',x))

In [103]:
products['wow'] = products['word_count'].apply(lambda x: fn_word_count('wow',x))

In [104]:
products['hate'] = products['word_count'].apply(lambda x: fn_word_count('hate',x))

In [105]:
products.head()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int
	awesome	int
	great	int
	fantastic	int
	amazing	int
	love	int
	horrible	int
	bad	int
	terrible	int
	awful	int
	wow	int
	hate	int

Rows: 10

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
| Baby Tracker&reg; - Daily ... | I love this journal and ou... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-----------+---------+-------+-----------+---------+
|           word_count          | sentiment | awesome | great | fantastic | amazing |
+-------------------------------+-----------+---------+-------+-----------+---------+
| {'and': 3, 'love': 1, 'it'... |     1     |    0    |   0   |     0     |    0    |
| {'and': 2, 'quilt': 1, 'it... |     1     |    0    |   0   |     0     |    0    |
| {'ingenious': 1, 'and': 3,... |     1     |    0    |   0   |     0     |    0    |
| {'and': 2, 'parents!!': 1,... |     1     |    0    |   1   |     0     |    0    |
| {'and': 2, 'this': 2, 'her... |     1     |    0    |   1   |     0     |    0    |
| {'shop': 1, 'noble': 1, 'i... |     1     |    0    |   0   |     0     |    0    |
| {'and': 2, 'all': 1, 'righ... |     1     |    0    |   0   |     0     |    0    |
| {'and': 1, 'help': 1, 'giv... |     1     |    0    |   0   |     0     |    0    |
| {'journal.': 1, 'nanny': 1... |     1     |    0    |   0   |     0     |    0    |
| {'all': 1, 'forget': 1, 'j... |     1     |    0    |   0   |     0     |    0    |
+-------------------------------+-----------+---------+-------+-----------+---------+
+------+----------+-----+----------+-------+-----+------+
| love | horrible | bad | terrible | awful | wow | hate |
+------+----------+-----+----------+-------+-----+------+
|  1   |    0     |  0  |    0     |   0   |  0  |  0   |
|  0   |    0     |  0  |    0     |   0   |  0  |  0   |
|  2   |    0     |  0  |    0     |   0   |  0  |  0   |
|  0   |    0     |  0  |    0     |   0   |  0  |  0   |
|  0   |    0     |  0  |    0     |   0   |  0  |  0   |
|  0   |    0     |  0  |    0     |   0   |  0  |  0   |
|  0   |    0     |  0  |    0     |   0   |  0  |  0   |
|  0   |    0     |  0  |    0     |   0   |  0  |  0   |
|  0   |    0     |  0  |    0     |   0   |  0  |  0   |
|  2   |    0     |  0  |    0     |   0   |  0  |  0   |
+------+----------+-----+----------+-------+-----+------+
[10 rows x 16 columns]

In [106]:
products['awesome'].sum()

2002

In [56]:
products['great'].sum()

42420

In [57]:
products['fantastic'].sum()

873

In [58]:
products['amazing'].sum()

1305

In [59]:
products['love'].sum()

40277

In [60]:
products['horrible'].sum()

659

In [61]:
products['bad'].sum()

3197

In [62]:
products['terrible'].sum()

673

In [63]:
products['awful'].sum()

345

In [64]:
products['wow'].sum()

131

In [65]:
products['hate'].sum()

1057

## "great" most used.  "wow" least used

In [107]:
train_data,test_data = products.random_split(.8, seed=0)

In [108]:
selected_words_model = graphlab.logistic_classifier.create(train_data,
                                                     target='sentiment',
                                                     features=selected_words,
                                                     validation_set=test_data)

PROGRESS: Logistic regression:
PROGRESS: --------------------------------------------------------
PROGRESS: Number of examples          : 133448
PROGRESS: Number of classes           : 2
PROGRESS: Number of feature columns   : 11
PROGRESS: Number of unpacked features : 11
PROGRESS: Number of coefficients    : 12
PROGRESS: Starting Newton Method
PROGRESS: --------------------------------------------------------
PROGRESS: +-----------+----------+--------------+-------------------+---------------------+
PROGRESS: | Iteration | Passes   | Elapsed Time | Training-accuracy | Validation-accuracy |
PROGRESS: +-----------+----------+--------------+-------------------+---------------------+
PROGRESS: | 1         | 2        | 0.232857     | 0.844299          | 0.842842            |
PROGRESS: | 2         | 3        | 0.390517     | 0.844186          | 0.842842            |
PROGRESS: | 3         | 4        | 0.539217     | 0.844276          | 0.843142            |
PROGRESS: | 4         | 5        |

In [109]:
selected_words_model['coefficients'].sort('value')

Columns:
	name	str
	index	str
	class	int
	value	float

Rows: 12

Data:
+-----------+-------+-------+------------------+
|    name   | index | class |      value       |
+-----------+-------+-------+------------------+
|  terrible |  None |   1   |  -2.09049998487  |
|  horrible |  None |   1   |  -1.99651800559  |
|   awful   |  None |   1   |  -1.76469955631  |
|    hate   |  None |   1   |  -1.40916406276  |
|    bad    |  None |   1   | -0.985827369929  |
|    wow    |  None |   1   | -0.0541450123333 |
|   great   |  None |   1   |  0.883937894898  |
| fantastic |  None |   1   |  0.891303090304  |
|  amazing  |  None |   1   |  0.892802422508  |
|  awesome  |  None |   1   |  1.05800888878   |
+-----------+-------+-------+------------------+
[12 rows x 4 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [110]:
selected_words_model['coefficients'].sort('value', ascending = False)

Columns:
	name	str
	index	str
	class	int
	value	float

Rows: 12

Data:
+-------------+-------+-------+------------------+
|     name    | index | class |      value       |
+-------------+-------+-------+------------------+
|     love    |  None |   1   |  1.39989834302   |
| (intercept) |  None |   1   |  1.36728315229   |
|   awesome   |  None |   1   |  1.05800888878   |
|   amazing   |  None |   1   |  0.892802422508  |
|  fantastic  |  None |   1   |  0.891303090304  |
|    great    |  None |   1   |  0.883937894898  |
|     wow     |  None |   1   | -0.0541450123333 |
|     bad     |  None |   1   | -0.985827369929  |
|     hate    |  None |   1   |  -1.40916406276  |
|    awful    |  None |   1   |  -1.76469955631  |
+-------------+-------+-------+------------------+
[12 rows x 4 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

## most positive - love
## most negative - terrible

In [111]:
selected_words_model.evaluate(test_data, metric='roc_curve')

{'roc_curve': Columns:
 	threshold	float
 	fpr	float
 	tpr	float
 	p	int
 	n	int
 
 Rows: 1001
 
 Data:
 +------------------+----------------+-------------------+-------+------+
 |    threshold     |      fpr       |        tpr        |   p   |  n   |
 +------------------+----------------+-------------------+-------+------+
 |       0.0        |      0.0       | 3.57079093019e-05 | 28005 | 5250 |
 | 0.0010000000475  |      1.0       |   0.999964292091  | 28005 | 5250 |
 | 0.00200000009499 | 0.99980952381  |   0.999928584181  | 28005 | 5250 |
 | 0.00300000002608 | 0.99980952381  |   0.999928584181  | 28005 | 5250 |
 | 0.00400000018999 | 0.999619047619 |   0.999892876272  | 28005 | 5250 |
 | 0.00499999988824 | 0.999619047619 |   0.999892876272  | 28005 | 5250 |
 | 0.00600000005215 | 0.999428571429 |   0.999857168363  | 28005 | 5250 |
 | 0.00700000021607 | 0.999428571429 |   0.999857168363  | 28005 | 5250 |
 | 0.00800000037998 | 0.999428571429 |   0.999857168363  | 28005 | 5250 |
 | 0.008

In [74]:
selected_words_model.show(view='Evaluation')

## What is the accuracy of the selected_words_model on the test_data?
accuracy = 0.8431419649291376

## What was the accuracy of the sentiment_model that we learned using all the word counts in the IPython Notebook above from the lectures? 
0.905

## What is the accuracy majority class classifier on this task? 
0.839

## How do you compare the different learned models with the baseline approach where we are just predicting the majority class?


In [112]:
diaper_champ_reviews = products[products['name'] == 'Baby Trend Diaper Champ']

In [113]:
diaper_champ_reviews['predicted_sentiment'] = sentiment_model.predict(diaper_champ_reviews, output_type='probability')

In [114]:
diaper_champ_reviews = diaper_champ_reviews.sort('predicted_sentiment', ascending=False)

In [115]:
diaper_champ_reviews.head()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int
	awesome	int
	great	int
	fantastic	int
	amazing	int
	love	int
	horrible	int
	bad	int
	terrible	int
	awful	int
	wow	int
	hate	int
	predicted_sentiment	float

Rows: 10

Data:
+-------------------------+-------------------------------+--------+
|           name          |             review            | rating |
+-------------------------+-------------------------------+--------+
| Baby Trend Diaper Champ | Baby Luke can turn a clean... |  5.0   |
| Baby Trend Diaper Champ | I LOOOVE this diaper pail!... |  5.0   |
| Baby Trend Diaper Champ | We researched all of the d... |  4.0   |
| Baby Trend Diaper Champ | My baby is now 8 months an... |  5.0   |
| Baby Trend Diaper Champ | This is absolutely, by far... |  5.0   |
| Baby Trend Diaper Champ | Diaper Champ or Diaper Gen... |  5.0   |
| Baby Trend Diaper Champ | Wow!  This is fabulous.  I... |  5.0   |
| Baby Trend Diaper Champ | I originally put this item... |  5.0   |
| Baby Trend Diaper Champ | Two girlfriends and two fa... |  5.0   |
| Baby Trend Diaper Champ | I am one of those super-cr... |  5.0   |
+-------------------------+-------------------------------+--------+
+-------------------------------+-----------+---------+-------+-----------+---------+
|           word_count          | sentiment | awesome | great | fantastic | amazing |
+-------------------------------+-----------+---------+-------+-----------+---------+
| {'all': 1, 'less': 1, "fri... |     1     |    0    |   0   |     0     |    0    |
| {'just': 1, 'over': 1, 'rw... |     1     |    0    |   0   |     0     |    0    |
| {'all': 2, 'just': 4, "don... |     1     |    0    |   0   |     0     |    0    |
| {"don't": 1, 'able': 2, 'o... |     1     |    0    |   2   |     0     |    0    |
| {'just': 3, 'money': 1, 's... |     1     |    0    |   0   |     0     |    0    |
| {'son': 2, 'all': 1, 'bags... |     1     |    0    |   0   |     0     |    0    |
| {'and': 4, 'this': 3, 'sti... |     1     |    0    |   0   |     0     |    0    |
| {'lysol': 1, 'all': 2, 'ba... |     1     |    0    |   0   |     0     |    0    |
| {'just': 1, '-': 3, 'both'... |     1     |    0    |   0   |     0     |    0    |
| {'all': 1, 'humid': 1, 'ju... |     1     |    0    |   0   |     0     |    0    |
+-------------------------------+-----------+---------+-------+-----------+---------+
+------+----------+-----+----------+-------+-----+------+---------------------+
| love | horrible | bad | terrible | awful | wow | hate | predicted_sentiment |
+------+----------+-----+----------+-------+-----+------+---------------------+
|  0   |    0     |  0  |    0     |   0   |  0  |  0   |    0.999999937267   |
|  1   |    0     |  0  |    0     |   0   |  0  |  0   |    0.999999917406   |
|  0   |    0     |  1  |    0     |   0   |  0  |  0   |    0.999999899509   |
|  0   |    0     |  1  |    0     |   0   |  0  |  0   |    0.999999836182   |
|  2   |    0     |  0  |    0     |   0   |  0  |  0   |    0.999999824745   |
|  0   |    0     |  0  |    0     |   0   |  0  |  0   |    0.999999759315   |
|  0   |    0     |  0  |    0     |   0   |  0  |  0   |    0.999999692111   |
|  0   |    0     |  0  |    0     |   0   |  0  |  0   |    0.999999642488   |
|  0   |    1     |  0  |    0     |   0   |  0  |  0   |    0.999999604504   |
|  1   |    0     |  0  |    0     |   0   |  0  |  0   |    0.999999486804   |
+------+----------+-----+----------+-------+-----+------+---------------------+
[10 rows x 17 columns]

## What is the ‘predicted_sentiment’ for the most positive review for ‘Baby Trend Diaper Champ’ according to the sentiment_model from the IPython Notebook from lecture?

0.999999937267

In [122]:
diaper_champ_reviews['predicted_sentiment'] = selected_words_model.predict(diaper_champ_reviews, output_type='probability')

In [124]:
diaper_champ_reviews = diaper_champ_reviews.sort('predicted_sentiment')

In [126]:
diaper_champ_reviews.tail()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int
	awesome	int
	great	int
	fantastic	int
	amazing	int
	love	int
	horrible	int
	bad	int
	terrible	int
	awful	int
	wow	int
	hate	int
	predicted_sentiment	float

Rows: 10

Data:
+-------------------------+-------------------------------+--------+
|           name          |             review            | rating |
+-------------------------+-------------------------------+--------+
| Baby Trend Diaper Champ | I have a 10 year old daugh... |  5.0   |
| Baby Trend Diaper Champ | This is absolutely, by far... |  5.0   |
| Baby Trend Diaper Champ | I've worked with kids more... |  5.0   |
| Baby Trend Diaper Champ | I love this diaper pail. I... |  4.0   |
| Baby Trend Diaper Champ | I love this diaper pail!  ... |  5.0   |
| Baby Trend Diaper Champ | Works great - no smells.  ... |  5.0   |
| Baby Trend Diaper Champ | I love this diaper pale an... |  5.0   |
| Baby Trend Diaper Champ | I received my Diaper Champ... |  5.0   |
| Baby Trend Diaper Champ | Love it, love it, love it!... |  5.0   |
| Baby Trend Diaper Champ | I LOVE LOVE LOVE this prod... |  4.0   |
+-------------------------+-------------------------------+--------+
+-------------------------------+-----------+---------+-------+-----------+---------+
|           word_count          | sentiment | awesome | great | fantastic | amazing |
+-------------------------------+-----------+---------+-------+-----------+---------+
| {'lysol': 1, 'all': 1, 'ju... |     1     |    0    |   0   |     0     |    0    |
| {'just': 3, 'money': 1, 's... |     1     |    0    |   0   |     0     |    0    |
| {'and': 4, 'genies': 1, 'a... |     1     |    0    |   0   |     0     |    0    |
| {'and': 1, 'old': 1, 'extr... |     1     |    0    |   0   |     0     |    0    |
| {'and': 3, 'love': 2, 'bei... |     1     |    0    |   0   |     0     |    0    |
| {'and': 2, 'bags.': 1, 'ga... |     1     |    0    |   2   |     0     |    0    |
| {'and': 3, 'love': 1, 'use... |     1     |    0    |   2   |     0     |    0    |
| {'bags.': 1, "don't": 1, '... |     1     |    0    |   0   |     0     |    0    |
| {'instead': 1, 'all': 1, '... |     1     |    0    |   0   |     0     |    0    |
| {'rating': 1, 'contacted':... |     1     |    0    |   1   |     0     |    0    |
+-------------------------------+-----------+---------+-------+-----------+---------+
+------+----------+-----+----------+-------+-----+------+---------------------+
| love | horrible | bad | terrible | awful | wow | hate | predicted_sentiment |
+------+----------+-----+----------+-------+-----+------+---------------------+
|  2   |    0     |  0  |    0     |   0   |  0  |  0   |    0.984739056527   |
|  2   |    0     |  0  |    0     |   0   |  0  |  0   |    0.984739056527   |
|  2   |    0     |  0  |    0     |   0   |  0  |  0   |    0.984739056527   |
|  2   |    0     |  0  |    0     |   0   |  0  |  0   |    0.984739056527   |
|  2   |    0     |  0  |    0     |   0   |  0  |  0   |    0.984739056527   |
|  1   |    0     |  0  |    0     |   0   |  0  |  0   |    0.989387539605   |
|  1   |    0     |  0  |    0     |   0   |  0  |  0   |    0.989387539605   |
|  3   |    0     |  0  |    0     |   0   |  0  |  0   |    0.996192539732   |
|  3   |    0     |  0  |    0     |   0   |  0  |  0   |    0.996192539732   |
|  3   |    0     |  0  |    0     |   0   |  0  |  0   |    0.998423414594   |
+------+----------+-----+----------+-------+-----+------+---------------------+
[10 rows x 17 columns]

## Why is the predicted_sentiment for the most positive review found using the model with all word counts (sentiment_model) much more positive than the one using only the selected_words (selected_words_model)? Hint: examine the text of this review, the extracted word counts for all words, and the word counts for each of the selected_words, and you will see what each model used to make its prediction.

In [120]:
diaper_champ_reviews[0]

{'amazing': 0,
 'awesome': 0,
 'awful': 0,
 'bad': 0,
 'fantastic': 0,
 'great': 1,
 'hate': 0,
 'horrible': 0,
 'love': 3,
 'name': 'Baby Trend Diaper Champ',
 'predicted_sentiment': 0.9984234145936198,
 'rating': 4.0,
 'review': 'I LOVE LOVE LOVE this product! It is SO much easier to use than the Diaper Genie, (you need a PHD in poopy to figure out how to use the darn thing!) and it even takes the same bags as my kitchen trash can, shich is super convenient, and cost efficient as I can buy them in bulk.The only reason for not rating it a 5 star was that I did have one small problem with it. The foam gasket in the barrell which keeps the poopy smell inside the unit ripped somehow, and it got VERY stinky. HOWEVER, I contacted the manufacturer though their website, and received an email back the same day stating that this was unusual, and that replacement gaskets were on their way to me. They arrived inside of a week and after replacing, it works great again! (They even sent me extras s